# APS106 - Lab 1

## Part 1 - Cartesian to polar coordinate conversion
For many problems in engineering and sciences, analysis can be simplified by converting between different coordinate systems. A common transformation is to convert 2-dimensional cartesian coordinates (x and y components) to polar coordinates (magnitude and phase angle). Fig 1 below illustrates how a vector can be defined using either coordinate system. Examples of situations where such transformations are commonly used include analysis of AC circuits, mechatronic control systems, ionic flow in fluids, and quantum mechanics.

![Fig 1](images/cart2polar.jpg "cartesian to polar")

Fig 1. Example of vector defined by both cartesian and polar coordinate systems where $r$ represents the magnitude and $\phi$ represents the phase angle.

The polar coordinates - magnitude $r$ and phase angle $\phi$ - can be computed from the cartesian $x$ and $y$ coordinates using the following equations:
$$
r = \sqrt{x^2 + y^2}
$$
$$
\phi = \tan^{-1}{\frac{y}{x}}
$$

In the following section, you will write functions that implement these equations and **return** polar coordinates given cartesian coordinates as parameters.

### Part 1.1 - Compute vector magnitude
Complete the function `magnitude` in the cell below. This function has two cartesian coordinate input **parameters** and **returns** the magnitude as defined using equation 1 above. Your function should return the result rounded to 3 decimal places. The table below defines the function inputs in greater detail.

| Parameter | Type | Description |
|-----------|------|-------------|
| `x` | `float` | The $x$ component of a two-dimensional vector in cartesian coordinates |
| `y` | `float` | The $y$ component of a two-dimensional vector in cartesian coordinates |

In [1]:
def magnitude(x, y):
    """
    (float, float) -> float

    Calculate and return the magnitude of the vector (x, y) rounded to 3 decimal places.

    Parameters
    ----------
    x : float
        The x-coordinate of the vector.
    y : float
        The y-coordinate of the vector.

    Returns
    -------
    float
        The magnitude of the vector (x, y).

    Examples
    --------
    >>> magnitude(3.0, 4.0)
    5.0
    """
    # To Do: Implement the function

    return round((x**2 + y**2)**0.5, 3)

#### Testing your function
Once you have written your function, you need to verify that it works correctly. Testing code involves generating sets of test inputs (called *test cases*), running the code with those inputs, and then verifying the outputs match the expected value. This week, we will walk you through some strategies for testing your functions. As we progress further into the semester and your programs become more complex, we will provide less explicit testing instructions and you will be responsible for testing your code. Developing good testing habits now will make you a better programmer and more efficient later in the course and in your career!

#### First test
Let's start with a simple test. Run the code in the cell below and check the result.

In [2]:
## Run a simple test
# Step 1: Define the input values
x = 3.0
y = 4.0
# Step 2: Call the function and store the result in a variable
r = magnitude(x, y)
# Step 3: Print the result
print("The function computed a magnitude of ", r)  # Expected output: 5.0

The function computed a magnitude of  5.0


#### It runs! =D ... or maybe it doesn't :/
If your function runs without error, great! Getting your code to run without crashing is the first step of verifying your code is correct. If your code crashed (you got an error message), you will need to correct the errors and then try running the test again.

#### The test runs without an error, now what?
The next step in the testing and debugging process is to verify the value that our function returned is actually correct. Take the values we passed into the function (*the arguments*) and plug them into the equation from above. Does the value returned by your function match this value? If not, congrats you've found a bug in your function. Correct the mistake and then run your test again. 

#### OK, I get the right value... I'm done right?
__Not yet!__ We have verified that our function works correctly in all possible *scenarios*. To understand what this means, let's consider an analogy. Imagine you're buying a mountain bike. You start shopping around and find a bike that has a great price. But after some research, you find that the company making the bike only tests their bike designs indoors on flat surfaces. So you have no idea how the bike will perform on a mountain trail! Would you trust and buy this bike? Best case the bike works fine but you might be nervous using it; worst case the bike falls apart, potentially injuring you or, at least, making you sad. You would be better off buying a bike that has been tested and evaluated for the conditions it is designed for.

What does this have to do with my function?... right now, our function is like the bike. We've done one test and it worked (like testing the bike indoors), but we need to think about the other scenarios the function will also need to work (like the bike being used outdoors on a mountain) so it doesn't fail and make us sad. 

#### Coming up with different test case scenarios
This doesn't mean choosing a bunch of random numbers and passing them to our function. Designing test cases is a core skill for any programmer (and is even a full time job for many engineers!). As your programs become more sophisticated later in the course, understanding how to effectively test your code will be very important and save you time and frustration. So practice and develop good habits around testing now and you will reap the rewards later on.

Let's look at another test case example.

In [3]:
# Test 2 - Make x negative (coordinates in the 2nd quadrant)
x = -3.0
y = 4.0
r = magnitude(x, y)
print("Test 2: The function computed a magnitude of ", r)  # Expected output: 5.0

Test 2: The function computed a magnitude of  5.0


Hopefully, your function still returns `5.0` for this test case. Even though we expect the same result from the function, we have diversified our test cases to include a negative value and increased our confidence that our function will work correctly for a wider range of inputs. Let's expand this even further with a more test cases.

In [4]:
## Test 3 - Coordinate in 3rd quadrant
x = -1.0
y = -1.0
r = magnitude(x, y)
print("Test 3: The function computed a magnitude of ", r)  # Expected output: 1.414

## Test 4 - Coordinate in 4th quadrant
x = 2.0
y = -10.0
r = magnitude(x, y)
print("Test 4: The function computed a magnitude of ", r)  # Expected output: 10.198

## Test 5 - Coordinate on the x-axis
x = 5.5
y = 0.0
r = magnitude(x, y)
print("Test 5: The function computed a magnitude of ", r)  # Expected output: 5.5

## Test 6 - Coordinate on the y-axis
x = 0.0
y = -7.15
r = magnitude(x, y)
print("Test 6: The function computed a magnitude of ", r)  # Expected output: 7.15

# Test 7 - Origin
x = 0.0
y = 0.0
r = magnitude(x, y)
print("Test 7: The function computed a magnitude of ", r)  # Expected output: 0.0

Test 3: The function computed a magnitude of  1.414
Test 4: The function computed a magnitude of  10.198
Test 5: The function computed a magnitude of  5.5
Test 6: The function computed a magnitude of  7.15
Test 7: The function computed a magnitude of  0.0


Notice how each of these tests represents a different input scenario (i.e., different coordinate quadrant, borders between quadrants, zero vector, etc.). Once your function returns the correct results for each of these test cases, you should have a reasonable degree of confidence that your fuction is correct and can move on to the next part.

### Part 2.2 - Compute the vector phase angle
Complete the `phase` function which computes and **returns** the phase angle of the vector in **radians** as defined in equation 2. The function uses the same input parameters as the `magntiude` function.

To complete this function, you will need to utilize the inverse tangent (arctangent) function from the `math` module. The documentation for the available trionometric functions can be found at the following link: https://docs.python.org/3/library/math.html#trigonometric-functions. You will notice that there are two different inverse tangent functions available within the `math` module: `atan` and `atan2`. Your function should use the function that will return an angle within the correct quadrant. Read the descriptions to select the appropriate function, then use the test cases below to verify that you made the correct selection. This is the beauty of designing and using test cases: you can always check and know whether your solution is correct by running your own code :).

In [5]:
import math 

def phase(x, y):
    """
    (float, float) -> float

    Calculate and return the phase of the vector (x, y) in radians rounded to 3 decimal places.

    Parameters
    ----------
    x : float
        The x-coordinate of the vector.
    y : float
        The y-coordinate of the vector.

    Returns
    -------
    float
        The phase of the vector (x, y) in radians.

    Examples
    --------
    >>> phase(3.0, 4.0)
    0.927
    """

    return round(math.atan2(y, x),3)

In [6]:
## Test the function with the same test cases as the magnitude function
# Test 1 - Quadrant 1
x = 3.0
y = 4.0
r = phase(x, y)
print("Test 1: The function computed a phase of ", r)  # Expected output: 0.927

# Test 2 - Quadrant 2
x = -3.0
y = 4.0
r = phase(x, y)
print("Test 2: The function computed a phase of ", r)  # Expected output: 2.214

# Test 3 - Quadrant 3
x = -1.0
y = -1.0
r = phase(x, y)
print("Test 3: The function computed a phase of ", r)  # Expected output: -2.356

# Test 4 - Quadrant 4
x = 2.0
y = -10.0
r = phase(x, y)
print("Test 4: The function computed a phase of ", r)  # Expected output: -1.373

# Test 5 - x-axis
x = 5.5
y = 0.0
r = phase(x, y)
print("Test 5: The function computed a phase of ", r)  # Expected output: 0.0

# Test 6 - y-axis
x = 0.0
y = -7.15
r = phase(x, y)
print("Test 6: The function computed a phase of ", r)  # Expected output: -1.571

# Test 7 - Origin
x = 0.0
y = 0.0
r = phase(x, y)
print("Test 7: The function computed a phase of ", r)  # Expected output: 0.0

Test 1: The function computed a phase of  0.9272952180016122
Test 2: The function computed a phase of  2.214297435588181
Test 3: The function computed a phase of  -2.356194490192345
Test 4: The function computed a phase of  -1.373400766945016
Test 5: The function computed a phase of  0.0
Test 6: The function computed a phase of  -1.5707963267948966
Test 7: The function computed a phase of  0.0


## Part 2 - Computing the position of a particle within an electrostatic precipitator

An electrostatic precipitator is a device that is typically used to capture and remove harmful particles from the exhaust of industrial and/or chemical processes. The device works in two stages:
1. Particles in the exhaust are charged (either positively or negatively) 
2. The particles pass through an air channel where high voltage electrodes attract the charged particles (see figure 1 below). The electrodes create an electric field that exerts a coulomb force on charged particles. The particles move towards the electrodes and get stuck to the electrodes, thus preventing the harmful particles from exiting with the exhaust.

Here, you will write a function to compute the vertical position of a particle as it passes through the precipitator and then design some test cases to verify that your function works correctly.

![Fig 2](images/esp.png)

Fig 2. Example of charged particle movement between high voltage electrodes. Particles enter through the opening on the left and are attracted to the oppositely charged electrode. In this example, a positively charged particle enters in the middle of the two electrodes and is accelerated towards the negatively charged electrode by the coulomb force, Fc. When the particle reaches the electrode, it sticks to the surface and stop moving.

If we assume this force is constant, the vertical position of the particle can be approximated using the following equation:
$$
p(t) = -\frac{1}{20000} \cdot \frac{qE}{m} t^2 + p_{\text{init}}
$$


| Variable     | Description                                           | Units                         |
|--------------|-------------------------------------------------------|-------------------------------|
| $q$          | Particle charge                                       | Nanocoulombs (nC)             |
| $E$          | Electric field strength between the electrodes        | Kilonewtons per coulomb (kN/C)|
| $m$          | Mass of the particle                                  | Nanograms (ng)                |
| $t$          | Time since the particle entered the air channel       | Microseconds (µs)             |
| $p_{\text{init}}$ | Initial vertical position of the particle      | Centimetres (cm)              |
| $p(t)$       | Vertical position of the particle at time `t`       | Centimetres (cm)              |

---


## Programming task
In the next cell, complete the function `compute_particle_height`. The function should calculate and *return* the vertical position of a particle within the electrostatic precipitator. The function accepts five input parameters:

- `q` – The charge of the particle in nanocoulombs
- `E` – The electric field strength in kilonewtons per coulomb
- `m` – The mass of the particle in nanograms
- `L` – The distance between the two electrodes in centimetres
- `t` – The time since the particle entered the air channel in microseconds

### Assumptions:
- The botton electrode is at position 0 cm and the top electrode is at position `L` cm. `L` will always be greater than zero.
- The bottom electrode will be negatively charged and the top will be positvely charged.
- Particles enter the channel halfway between the two plates (i.e., $p_{\text{init}} = L/2$ cm).
- Particles have zero vertical velocity when entering the air channel.
- __Particles immediately stop moving when contacting one of the electrodes.__
- All other forces (gravity, drag, etc.) are negligible.

Your function should round the result to **3 decimal places**.

**Hint**: This problem is designed to be solved without the use of conditional (if) statements (we'll get to those in a couple of weeks). Instead, for this lab, you may find the built-in Python functions `max` and `min` helpful. Try running
`help(min)` and 
`help(max)`
in the cell below to find out more about these functions.


In [13]:
def compute_particle_height(q,E,m,t,L):
    """
    (float, float, float, float, float) -> float
    
    Calculate the vertical height of a charged particle within an electrostatic precipitator.

    Parameters
    ----------
    q : float
        Charge of the particle in nanocoulombs (nC).
    E : float
        Electric field strength in kilonewtons per coulomb (kN/C).
    m : float
        Mass of the particle in nanograms (ng).
    t : float
        Time since the particle entered the precipitator in microseconds (µs).
    L : float
        Distance between the parallel plate electrodes in centimetres (cm).

    Returns
    -------
    float
        The height of the particle in centimetres (cm).

    Examples
    --------
    >>> compute_particle_height(0, 150, 9.2, 3.6, 5.0)
    2.5

    >>> compute_particle_height(2.3, 150, 9.2, 26.8, 5.0)
    1.153

    >>> compute_particle_height(-2.3, 160, 9.2, 36.8, 5.0)
    5.0
    """
    # To Do: Implement the function
    return round(min(max(0.0, (-1.0/20000) * (q*E)/(m) * (t**2) + (L/2)), L * 1.0),3)

## Testing your function
As before, you will need to test your function to verify its correctness. Let's start with a simple test.

In [14]:
## Run a simple test
# step 1: execute the function and assign the return value to a variable
particle_position = compute_particle_height(2.3, 150, 9.2, 26.8, 5.0)
# step 2: display the result
print("The function returned: ", particle_position)

The function returned:  1.153


#### It runs! =D ... or maybe it doesn't :/
If your code crashed (you got an error message), you will need to correct the errors and then try running the test again.

Once your function runs without error and returns the correct value, we can move on to designing more test cases.

In [11]:
## Test #2 - Larger particle
# Double the mass of the particle
# step 1: execute the function and assign the return value to a variable
particle_position = compute_particle_height(2.3, 150, 18.4, 26.8, 5.0)
# step 2: display the result
print("The function returned: ", particle_position)

The function returned:  1.827


Hopefully, your function returns `1.827`. You can verify that is the correct value by plugging the values into the equation above. If your function is returning the correct value, great! But should we feel more confident that our function is correct? Unfortunately, not really. Even though we changed the input and got a different result, all we're testing is whether our function implemented the equation correctly. We have not tested the other components of our function. Let's look at another example to understand what we mean. This time, let's increase the time parameter significantly to `260.8`.

In [12]:
## Test #2 - Larger particle
# Increase the time
# step 1: execute the function and assign the return value to a variable
particle_position = compute_particle_height(2.3, 150, 18.4, 260.8, 5.0)
# step 2: display the result
print("The function returned: ", particle_position)

The function returned:  0


If we plug this into our equation we get:
$$
p(t) = -\frac{1}{20000} \cdot \frac{qE}{m} t^2 + p_{\text{init}} = -\frac{1}{20000} \cdot \frac{2.3\cdot150}{18.4}\cdot260.8^2 + 2.5 = -61.266
$$

But, one of the assumptions for our function was: "__Particles immediately stop moving when contacting one of the electrodes__". This means that our positive particle from this test stops moving when it hits the bottom electrode and the height cannot be lower than zero. So, the value returned by our function should be `0.0` and not `-61.266`.

This test case represents a different scenario because, unlike before, the output is not simply the result of plugging values into our equation. Instead, it repesents a scenario where the particle hits one of the boundaries. Identifying these different scenarios is the key to effectively testing your code.

#### Other scenarios
Some other scenarios that we could consider for this function include:
 - The charge of the particle (does the particle move towards the top boundary when a negative charge is passed to the function? what if the particle has zero charge?)
 - Does the function return the correct height when the particle makes contact with the top electrode?
 
 
Now design some of your own test cases, and try them out in the cell below.

In [ ]:
# Write your own test cases here



Once you have thoroughly tested your code, submit this notebook to Gradescope.